parser.go

```diff
@@ -900,35 +900,45 @@ func (p *parser) parseParamComment(pkgPath, pkgName string, operation *Operation
 		}
 	}
 
-	if strings.HasPrefix(goType, "[]") || strings.HasPrefix(goType, "map[]") || goType == "time.Time" {
-		schema, err := p.parseSchemaObject(pkgPath, pkgName, goType, true)
-		if err != nil {
-			p.debug("parseResponseComment cannot parse goType", goType)
-		}
-		operation.RequestBody.Content[ContentTypeJson] = &MediaTypeObject{
-			Schema: *schema,
-		}
-	} else {
-		typeName, err := p.registerType(pkgPath, pkgName, matches[3])
+	s, err := p.parseBodyType(pkgPath, pkgName, goType)
+	if err != nil {
+		return err
+	}
+	operation.RequestBody.Content[ContentTypeJson] = &MediaTypeObject{
+		Schema: *s,
+	}
+
+	return nil
+}
+
+func (p *parser) parseBodyType(pkgPath, pkgName, typeName string) (*SchemaObject, error) {
+	if strings.HasPrefix(typeName, "[]") || strings.HasPrefix(typeName, "map[]") || typeName == "time.Time" {
+		schema, err := p.parseSchemaObject(pkgPath, pkgName, typeName, true)
 		if err != nil {
-			return err
-		}
-		if isBasicGoType(typeName) {
-			operation.RequestBody.Content[ContentTypeJson] = &MediaTypeObject{
-				Schema: SchemaObject{
-					Type: &stringType,
-				},
-			}
-		} else {
-			operation.RequestBody.Content[ContentTypeJson] = &MediaTypeObject{
-				Schema: SchemaObject{
-					Ref: addSchemaRefLinkPrefix(typeName),
-				},
-			}
+			p.debug("parseResponseComment cannot parse type", typeName)
 		}
+		return schema, nil
 	}
 
-	return nil
+	// handle oneOf/anyOf/allOf/not
+	sob, err := p.handleCompositeType(pkgPath, pkgName, typeName)
+	if sob != nil || err != nil {
```

Comment:
this isn't a common convention that I've seen, it's usually error handling first, then if that's nil handle the next part but I'm guessing it works because the code calling it has the error handler checked before the SchemaObject?

It could also just be something I haven't run into yet.

Reply:
That's right.

It's an alternative to this pattern, where you check a condition before calling a particular handler,

```
if shouldUseFooHandler {
  result, err := p.fooHandler(...)
  if (err) {
    return nil, err
  }
  return result, nil
} // else need to use a different handler
```

Where instead you move the condition check into the handler, and perform that check internally, and always call the handler, relying on the result to tell you if it a) handled it, b) error'd or c) did not handle it and did not error

```
result, err := p.fooHandler(...)
if result != nil || err != nil {
  return result, err
}

// try to use a different handler
```

I'm coming to understand that cleverness is highly discouraged in Go, so if it's a little jarring I can refactor from the latter to the former.

Reply:
The alternative would look something like this if you like it better,

```
	isCompoundType, _ := regexp.MatchString("(oneOf|anyOf|allOf|not)\\([^\\)]*\\)", typeName)
	if isCompoundType {
		sob, err := p.handleCompoundType(pkgPath, pkgName, typeName)
		if err != nil {
			return nil, err
		}
		return sob, nil
	}
```